woocommerce/woocommerce
# Abort update and show a notification if CREATE/ALTER TABLE fails during install or update

### All Submissions:


### Changes proposed in this Pull Request:

The original task was about simply displaying an error notice whenever a table creation failed during install/update, however the following additional changes are introduced:

* If any CREATE TABLE or ALTER TABLE database query fails, skip all the remaining queries, including the update functions. This is to prevent further failures, since the database schema won't be in a predictable state.
* Additionally to skipping the remaining queries, don't bump the WooCommerce and database versions. This is to allow retrying the pending updates at a later time.

Closes #26311.

### How to test the changes in this Pull Request:

**1.** Take note of your current WooCommerce and database versions since you'll mess with them (you can execute `wp option get woocommerce_version` and `wp option get woocommerce_db_version` in a terminal to get them).

**2.** Execute the following database commands to create a user that doesn't have privileges to perform database schema changes:

```
CREATE USER 'theuser'@'%' IDENTIFIED BY 'thepassword';
GRANT SELECT,UPDATE,INSERT,DELETE ON *.* TO 'theuser';
```
(you can execute `SHOW GRANTS FOR 'theuser'` to verify that the user got created correctly)

**3.** Modify your `wp-config.php` so that the values for `DB_USER` and `DB_PASSWORD` correspond to the user you just created (so `'theuser'` and `'thepassword'`, respectively).

**4.** Modify your `class-woocommerce.php` so that the `$version` property has a value that is higher than the current one, e.g. `"10.0"`.

**5.** Open `class-wp-install.php` and all the following to the SQL that the `get_schema` method returns:

```
CREATE TABLE foobar (id INT);
CREATE TABLE fizzbuzz (id INT);
```

**6.** Browse your WP admin area and you'll see an error notice like this one:

![image](https://user-images.githubusercontent.com/937723/82347497-faaf8480-99f7-11ea-8767-2f9bebc658c2.png)

**7.** Reload the page and verify that the notice is still here.

**8.** Verify that neither the "foobar" nor the "fizzbuzz" tables have been created, you can do that by running `wp db query "show tables like 'f%'"` (it should run without displaying anything).

**9.** Verify that the WooCommerce and database versions haven't changed (repeat step 1 and you should get the same values).

**10.** Restore the original user and password in `wp_config.php`.

**11.** Reload your admin page. The notice should have disappeared.

**12.** Repeat step 8, now the tables should exist.

**13.** Repeat step 1, now both versions should be `10.0`.

You're done testing! Now it's cleanup time:

**14.** Restore the original WooCommerce version in `class-woocommerce.php`.

**15.** Restore the WooCommerce and database version numbers by running `wp option set woocommerce_version <original version>` in your terminal, and same for `woocommerce_db_version`.

**16.** Drop the dummy tables, you can do that by running `wp db query "drop table foobar"` and `wp db query "drop table fizzbuzz"` in your terminal.

**17.** Unless you want to keep it for future tests, remove the user you created, you can do that by running `wp db query "drop user 'theuser'"` in your terminal.

### Other information:


### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.
